subxt/src/backend/rpc/reconnecting_rpc_client/mod.rs

```diff
@@ -0,0 +1,630 @@
+#[error("The connection was closed because of `{0:?}` and reconnect initiated")]
+pub struct DisconnectedWillReconnect(String);
+
+/// Serialized JSON-RPC params.
+#[derive(Debug, Clone)]
+pub struct RpcParams(Option<Box<RawValue>>);
+
+impl RpcParams {
+    /// Create new [`RpcParams`] from JSON.
+    pub fn new(json: Option<Box<RawValue>>) -> Self {
+        Self(json)
+    }
+}
+
+impl ToRpcParams for RpcParams {
+    fn to_rpc_params(self) -> Result<Option<Box<RawValue>>, serde_json::Error> {
+        Ok(self.0)
+    }
+}
+
+#[derive(Debug)]
+enum Op {
+    Call {
+        method: String,
+        params: RpcParams,
+        send_back: oneshot::Sender<MethodResult>,
+    },
+    Subscription {
+        subscribe_method: String,
+        params: RpcParams,
+        unsubscribe_method: String,
+        send_back: oneshot::Sender<Result<Subscription, Error>>,
+    },
+}
+
+/// Error that can occur when for a RPC call or subscription.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// The client was dropped by the user.
+    #[error("The client was dropped")]
+    Dropped,
+    /// The connection was closed and reconnect initiated.
+    #[error(transparent)]
+    DisconnectedWillReconnect(#[from] DisconnectedWillReconnect),
+    /// Other rpc error.
+    #[error("{0}")]
+    RpcError(RpcError),
+}
+
+/// Represent a single subscription.
+pub struct Subscription {
+    id: SubscriptionId<'static>,
+    stream: mpsc::UnboundedReceiver<SubscriptionResult>,
+}
+
+impl Subscription {
+    /// Returns the next notification from the stream.
+    /// This may return `None` if the subscription has been terminated,
+    /// which may happen if the channel becomes full or is dropped.
+    ///
+    /// **Note:** This has an identical signature to the [`StreamExt::next`]
+    /// method (and delegates to that). Import [`StreamExt`] if you'd like
+    /// access to other stream combinator methods.
+    #[allow(clippy::should_implement_trait)]
+    pub async fn next(&mut self) -> Option<SubscriptionResult> {
+        StreamExt::next(self).await
+    }
+
+    /// Get the subscription ID.
+    pub fn id(&self) -> SubscriptionId<'static> {
+        self.id.clone()
+    }
+}
+
+impl Stream for Subscription {
+    type Item = SubscriptionResult;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+    ) -> task::Poll<Option<Self::Item>> {
+        match self.stream.poll_recv(cx) {
+            Poll::Ready(Some(msg)) => Poll::Ready(Some(msg)),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl std::fmt::Debug for Subscription {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("Subscription {:?}", self.id))
+    }
+}
+
+/// JSON-RPC client that reconnects automatically and may loose
+/// subscription notifications when it reconnects.
+#[derive(Clone, Debug)]
+pub struct Client {
```

Comment:
Nit: I wonder whether to call this `RpcClient` to align with the same name in other places.